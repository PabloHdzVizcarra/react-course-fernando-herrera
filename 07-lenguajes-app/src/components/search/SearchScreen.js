import React, { useMemo } from 'react';
import queryString from 'query-string';
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router-dom';
import { getHeroByName } from '../../selectors/getHeroByNmae';


export const SearchScreen = ({history}) => {
  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);
  const [{ searchText }, handleInputChange] = useForm({ searchText: q });
  
  const handleSearch = (event) => {
    event.preventDefault();

    history.push(`?q=${searchText}`);
  }

  const heroesFilter = useMemo(() => getHeroByName(q), [q]);

  return (
    <div>
      <h1>Search</h1>
      <hr />
      
      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr />
          <form
            onSubmit={handleSearch}
          >
            <input
              autoComplete='off'
              type="text" 
              placeholder="busca tu heroe"
              className='form-control'
              name='searchText'
              value={searchText}
              onChange={handleInputChange}
            />
            
            <button
              type='submit'
              className='btn mt-1 btn-block btn-outline-primary'
            >
              Buscar
            </button>
          </form>
          
        </div>
        <div className="col-7">
          <h4>Resultados</h4>
          <hr />

          {
            (q === '')
              && 
              <div className='alert alert-info'>
                Busca un Heroe
              </div>
          }

          {
            (q !== '' && heroesFilter.length === 0)
              && 
              <div className='alert alert-danger'>
                No existen heroes {q}
              </div>
          }
          
          {
            heroesFilter.map((hero) => (
              <HeroCard 
                key={hero.id}
                {...hero}
              />
            ))
          }
        </div>
      </div>
    </div>
  )
}
