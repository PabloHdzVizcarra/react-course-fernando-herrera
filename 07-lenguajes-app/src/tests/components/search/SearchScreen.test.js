import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';


describe('test en <SearchScreen />', () => {
  
  test('should show correctly with default values', () => {
    
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']} >
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Busca un Heroe');

  });

  test('should mostrar a batman y el input con el calor del queryString', () => {
    
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']} >
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper).toMatchSnapshot();
  });

  test('should show text error when not exists hero', () => {

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']} >
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-danger').text().trim()).toBe('No existen heroes batman123');
  });

  test('should llamar el push de history', () => {
    
    const history = {
      push: jest.fn(),
    }

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']} >
        <Route
          path="/search"
          component={() => <SearchScreen history={history} />}
        />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
      target: {
        name: 'searchText',
        value: 'batman'
      }
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault() { }
    });

    expect(history.push).toHaveBeenCalledWith('?q=batman');

  })
  
  
  
  
})
