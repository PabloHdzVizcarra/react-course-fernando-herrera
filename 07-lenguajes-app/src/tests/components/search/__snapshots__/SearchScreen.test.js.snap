// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test en <SearchScreen /> should mostrar a batman y el input con el calor del queryString 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="busca tu heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="card ms-3 animate__animated animate__fadeIn"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test en <SearchScreen /> should show correctly with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="busca tu heroe"
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Busca un Heroe
      </div>
    </div>
  </div>
</div>
`;

exports[`test en <SearchScreen /> should show text error when not exists hero 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="busca tu heroe"
          type="text"
          value="batman123"
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No existen heroes 
        batman123
      </div>
    </div>
  </div>
</div>
`;
